shared/src/question-generators/propositional-logic/minimize.ts

```diff
@@ -0,0 +1,189 @@
+
+    const typingsAids = getTypingAids(lang)
+    const question: FreeTextQuestion = {
+      type: "FreeTextQuestion",
+      name: MinimizePropositionalLogic.name(lang),
+      path,
+      prompt: "$\\varPhi^*=$",
+      text: t(translations, lang, "text", [
+        randExpression.toString(true),
+        t(translations, lang, variant),
+      ]),
+      feedback: feedbackFunction(randExpression, variant, lang),
+      checkFormat: checkFormat(varNames),
+      typingAid: [
+        typingsAids.leftParenthesis,
+        typingsAids.rightParenthesis,
+        typingsAids.logicalOr,
+        typingsAids.logicalAnd,
+        typingsAids.logicalNot,
+      ].concat(getTypingAidsVars(lang, randExpression.getVariableNames()).typingAidVars),
+    }
+
+    return { question }
+  },
+}
+
+/**
+ * Simple function to parse the user input
+ * Catch ParsingError and show, it's not parsable
+ * @param varNames
+ */
+function checkFormat(varNames: string[]): FreeTextFormatFunction {
+  return ({ text }) => {
+    const res = PropositionalLogicParser.parse(text)
+
+    return {
+      valid: !(res instanceof ParserError || _.difference(res.getVariableNames(), varNames).length > 0),
+      message: `$ ${tokenToLatex(text)}$`,
+    }
+  }
+}
+
+/**
+ * Feedback function to check if the user input is
+ * - a boolean expression
+ * - either a DNF or CNF
+ * - equivalent to solutionExpression
+ * - has the same size as the minDNF/CNF of solutionExpression
+ *
+ * Todo: currently only **DNF** is supported
+ *
+ * @param solutionExpression - base expression
+ * @param functionType - normal form type
+ * @param lang
+ */
+function feedbackFunction(
+  solutionExpression: SyntaxTreeNodeType,
+  functionType: "DNF", // | "CNF"
+  lang: "en" | "de",
+): FreeTextFeedbackFunction {
+  return ({ text }) => {
+    const minExpressionDNF: SyntaxTreeNodeType = minimizeExpressionDNF(solutionExpression)
+    const correctAnswer = "$" + minExpressionDNF.toString(true) + "$"
+
+    const userExpression = PropositionalLogicParser.parse(text)
+    if (userExpression instanceof ParserError) {
+      return {
+        correct: false,
+        feedbackText: t(translations, lang, "freetext_feedback_parse_error"),
+        correctAnswer,
+      }
+    }
+
+    const isCorrectForm = functionType === "DNF" ? userExpression.isDNF() : userExpression.isCNF()
+    if (!isCorrectForm) {
+      return {
+        correct: false,
+        feedbackText: t(translations, lang, "freetext_feedback_no_normal_form", [
+          t(translations, lang, functionType),
+        ]),
+        correctAnswer,
+      }
+    }
+
+    if (!compareExpressions([minExpressionDNF, userExpression])) {
+      return {
+        correct: false,
+        feedbackText: t(translations, lang, "freetext_feedback_not_equivalent"),
+        correctAnswer,
+      }
+    }
+
+    if (minExpressionDNF.getLiteralSize() !== userExpression.getLiteralSize()) {
```

Comment:
Due to the number of literals determining minimality, redundant negations currently aren't considered (i.e. `A` and `\not(\not A)` are considered equally minimal)

Reply:
Yes I understand what you mean and have thought about it. But I currently don't know how to compare two expression which are equal and minimal, but have different amount of negations. 

Do you have an idea how to check if two expressions have the same length?

Reply:
Since the comparison to detect equivalent expressions works, it would probably suffice to check for redundant negations (maybe like this `this.negated && this.negate().negated`) and give the feedback that the solution could be minimized further.

Reply:
You could check `userExpression.simplifyNegation().toString().length < userExpression.toString().length` (or something like that). simplifyNegation moves all negations to the leaves.

Reply:
<img width="892" alt="Screenshot-2024-11-12-at-15 05 37" src="https://github.com/user-attachments/assets/9c827c48-8057-4c21-8bca-d4c10f156e32">
And unfortunately the Parser already parses `\not ( \not a)` to `a`. So both ideas would not work

Reply:
Ah, I remember. I implemented `not` as a flag, not as an operator. You could remove unnecessary spaces from the input and just use it directly for comparison. Or we ignore this case?

Reply:
Just using the input string would not work either, since (not sure) there could be two different dnfs, which are equal and have the same length, but a different amount of `\not`'s. Then comparison of `.toString().length()` would not work 

For now I would ignore this case, maybe we find a solution in the future